separatedunit.cpp

```diff
@@ -0,0 +1,1286 @@
+        Cohere(*(toCopy.coherentUnits[i]));
+    }
+}
+
+void SeparatedUnit::Decohere(bitLenInt start, bitLenInt length, CoherentUnit& destination)
+{
+    DecohereOrDispose(true, start, length, &destination);
+}
+
+void SeparatedUnit::Dispose(bitLenInt start, bitLenInt length) { DecohereOrDispose(false, start, length, NULL); }
+
+/// PSEUDO-QUANTUM Direct measure of bit probability to be in |1> state
+double SeparatedUnit::Prob(bitLenInt qubitIndex)
+{
+    QbLookup qbl = qubitLookup[qubitIndex];
+    return coherentUnits[qbl.cu]->Prob(qbl.qb);
+}
+
+double SeparatedUnit::ProbAll(bitCapInt perm)
+{
+    bitLenInt i, j;
+    bitLenInt cuLen, qb;
+    bitCapInt partPerm;
+    double result = 1.0;
+
+    for (i = 0; i < coherentUnits.size(); i++) {
+        cuLen = coherentUnits[i]->GetQubitCount();
+        partPerm = 0;
+        for (j = 0; j < cuLen; j++) {
+            qb = qubitInverseLookup[i * qubitCount + j];
+            partPerm |= (perm & (1 << qb)) >> (qb - j);
+        }
+        result *= coherentUnits[i]->ProbAll(partPerm);
+    }
+
+    return result;
+}
+
+void SeparatedUnit::ProbArray(double* probArray)
+{
+    bitLenInt i;
+
+    for (i = 0; i < qubitCount; i++) {
+        probArray[i] = Prob(i);
+    }
+}
+
+/// Measure a bit
+bool SeparatedUnit::M(bitLenInt qubitIndex)
+{
+    bool result;
+    QbLookup qbl = qubitLookup[qubitIndex];
+    result = coherentUnits[qbl.cu]->M(qbl.qb);
+
+    bitLenInt qbCount = coherentUnits[qbl.cu]->GetQubitCount();
+    if (qbCount > 1) {
+        std::shared_ptr<CoherentUnit> ncu(new CoherentUnit(1, 0, rand_generator_ptr));
+        coherentUnits[qbl.cu]->Decohere(qbl.qb, 1, *ncu);
+        coherentUnits.push_back(ncu);
+
+        bitLenInt i;
+        bitLenInt invLookup = qubitInverseLookup[qbl.cu * qubitCount + qbl.qb];
+        for (i = qubitLookup[invLookup].qb; i < (qbCount - 1); i++) {
+            qubitLookup[qubitInverseLookup[qbl.cu * qubitCount + i]].qb--;
+            qubitInverseLookup[qbl.cu * qubitCount + i] = qubitInverseLookup[qbl.cu * qubitCount + i + 1];
+        }
+        qubitLookup[invLookup].cu = coherentUnits.size() - 1;
+        qubitLookup[invLookup].qb = 0;
+        qubitInverseLookup[(coherentUnits.size() - 1) * qubitCount] = invLookup;
+    }
+
+    return result;
+}
+
+/// Measure permutation state of a register
+bitCapInt SeparatedUnit::MReg(bitLenInt start, bitLenInt length)
+{
+    bitCapInt result = 0;
+    bitLenInt i, j;
+    bitLenInt qbCount, invLookup;
+    QbListEntry qbe;
+    QbLookup qbl;
+
+    std::vector<QbListEntry> qbList(length);
+    GetOrderedBitList(start, length, qbList);
+
+    j = 0;
+    for (i = 0; i < qbList.size(); i++) {
+        qbe = qbList[i];
+        result |= (coherentUnits[qbe.cu]->MReg(qbe.start, qbe.length)) << j;
+        j += qbe.length;
+    }
+
+    for (i = 0; i < length; i++) {
+        qbl = qubitLookup[start + i];
+        qbCount = coherentUnits[qbl.cu]->GetQubitCount();
+        if (qbCount > 1) {
```

Comment:
This code looks almost identical to `M()`.  If so, or if it's close enough, can this be reduced to bunch of calls to `M()`?

Practically speaking, though, it seems like once you `M()` anything that intersects with any of these CUs, they should all be measured to some explicit state and decomposed to individual CUs en mass.  I'm not entirely sure that is what is happening here.

Reply:
While the two are logically equivalently, ```MReg()``` is much more preferable for performance reasons, like reducing iterations by a factor of 256 for an 8 bit register in worst case, and still giving significant gains in better cases.

Yes, it's much harder to understand. I think it's worth saving a multiplicative factor of 256 iterations, unfortunately. A lot of the code in SeparatedUnit is driven by this consideration, using optimized register-like methods to their fullest at every opportunity. In all those cases, for 8 bits registers, it will usually reduce iterations by a multiplicative factor of 256 in worst case. In fact, for speed reasons, we should go the other direction with Pauli operators, mirroring this approach, and again save a factor of 256.

The reason it's particularly complicated is because of arbitrary initial entanglement. CUs might contain both bits involved in the MReg and bits not involved in the MReg. We quick sort bit indices on entanglement in order to maximize the potential for register-wise method application. I think this is a necessary evil.